Tensors/src/main/scala/com/thoughtworks/compute/Tensors.scala

```
@@ -1125,14 +1125,55 @@ trait Tensors extends OpenCL {
     }
 
     /** Returns an asynchronous task to read this [[Tensor]] into a [[scala.Array]],
-      * which is linearized in row-majoy order.
+      * which is linearized in row-major order.
       *
       * @group slow
       */
     def flatArray: Future[Array[closure.JvmValue]] = {
       flatBuffer.intransitiveMap(closure.valueType.memory.toArray).run
     }
 
+    /** Recursive function to reshape a flat array into a multi-dimensional one
+      */
+    private[Tensors] def reshapeArray(array:Array[_], shape:Array[Int]): Array[_] = {
```

When `shape` is `Array.empty`, the tensor is a scalar value, not an `Array`.

This method should either throw an exception for scalar value, or change the return type.

---

Right! If scalars are considered distinct from 1d arrays, then I think it makes sense that the function `readArray` would throw an exception.

---

If scalars are considered distinct from 1d arrays, why 1D arrays and 2D array shares the same method? Considering the `Array[_]` return type is not actually usable without an `anInstanceOf`.

There are three options:

1. Create two methods for scalars and 1+ dimensional arrays, respectively
2. Create one `Any` methods for both scalars and 1+ dimensional arrays
3. Create n methods for different rank of tensors, respectively

What do you think?